src/server/master_server/master_metadata_service_impl.cc

```diff
@@ -201,6 +205,14 @@ grpc::Status MasterMetadataServiceImpl::HandleFileChunkRead(
 
   // Set the file chunk metadata in reply
   *reply->mutable_metadata() = file_chunk_metadata_or.ValueOrDie();
```

Comment:
We are setting the metadata here which contains the previously selected primary. We need to check that the previous primary is in the set of returned chunk locations below. If the primary isn't in the set of returned locations, it means it is down and should select a new primary from the returned locations. Otherwise operations will always fail since we keep returning the same primary but the primary is down.